### ISSUE TITLE
Segmentation fault: 11 trying to install on device

### ISSUE THREAD

**Issue by tvolkert** [libimobiledevice/ideviceinstaller#73]
Trying to install an IPA onto my iOS 10.3.1 device, I get the following output, and then the segfault:

```
$ ideviceinstaller -i Payload/hello_world.app/
Uploading hello_world.app package contents... DONE.
Installing '(null)'
Install: CreatingStagingDirectory (5%)
Install: ExtractingPackage (15%)
Install: InspectingPackage (20%)
Install: TakingInstallLock (20%)
Install: PreflightingApplication (30%)
Install: InstallingEmbeddedProfile (30%)
Install: VerifyingApplication (40%)
Install: CreatingContainer (50%)
Install: InstallingApplication (60%)
Install: PostflightingApplication (70%)
Install: SandboxingApplication (80%)
Install: GeneratingApplicationMap (90%)
Segmentation fault: 11
```

The strange thing is that the app *does* get successfully installed and runs ok.

I've installed the latest `ideviceinstaller` from HEAD (as of today).

Crash report is available at https://gist.github.com/tvolkert/217237b40a626e724a25b02b619a233d

**Comment by tvolkert** [comment]
I've also verified that this crash only happens in newer iOS versions.  Tested against iOS 10.1.1, and everything worked.  Tested against iOS 10.3.1, and I get the crash.

I've uploaded an IPA that demonstrates this behavior to http://54.91.18.239/bugs/ideviceinstaller/73/hello_flutter.ipa

**Comment by stevenyix** [comment]
i'm experiencing an identical issue with the same output. attempting to deploy an app on an iphone 6 running ios 10.3.1

**Comment by codemonkey1337** [comment]
I'm getting this too, iOS 10.3.1, macOS Sierra 10.12.4 with -HEAD install.

**Comment by foobar8675** [comment]
Same here. Are there any work arounds for this?

**Comment by jpollack** [comment]
I'm not familiar with this codebase, so, don't have any understanding of the deeper implications but I ran into this bug so made a fix for this issue.  In ideviceinstaller.c::status_cb, strcmp and printf are passed char *, without checking first if the char * are NULL.  Adding a trivial check for NULL resolves it.

https://github.com/libimobiledevice/ideviceinstaller/pull/75

**Comment by foobar8675** [comment]
Thanks @jpollack . I've just gone back to launching via xcode for now, which is fine for a temporary work around for launching RN apps to an ios device.

**Comment by crummy** [comment]
ios-deploy still works on 10.3 if you're on Mac and need a workaround: https://github.com/phonegap/ios-deploy

**Comment by JonGabilondoAngulo** [comment]
Hi,

I'm  considering the following fix. 
The problem is that in the last message there is no message  "Complete":
We can see inthe the status plist that there is no information. Apple messed up in 10.3.
                 <?xml version="1.0" encoding="UTF-8"?>
                 <!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
                 <plist version="1.0">
                 <dict>
                 <key>CFBundleIdentifier</key>
                 <string>com.hp.hmc.UICatalog8</string>
                 </dict>
                 </plist>
                 
              
A solution would be that if is last status step, there is no error, and there is no Complete message, add the "Complete" message.

What are the pitfalls of this solution in your opinion ? 
Last step of status .. not being error .. not having status .. and iOS > 10.3 .. 
It's obviously not 100% perfect, but is it fair to think that if there was an error we would get it ?


static void status_cb(plist_t command, plist_t status, void *unused)
{
	if (command && status) {
		char* command_name = NULL;
		instproxy_command_get_name(command, &command_name);

		/* get status */
		char *status_name = NULL;
		instproxy_status_get_name(status, &status_name);

		if (status_name) {
			if (!strcmp(status_name, "Complete")) {
				command_completed = 1;
			}
        }

		/* get error if any */
		char* error_name = NULL;
		char* error_description = NULL;
		uint64_t error_code = 0;
		instproxy_status_get_error(status, &error_name, &error_description, &error_code);

		/* output/handling */
		if (!error_name) {
            
            if (!status_name) {
                // FIX. If no status name then we consider is a Complete. It happens in ios 10.3.
                status_name = strdup("Complete");
                command_completed = 1;
                
                
            }
            
			if (!strcmp(command_name, "Browse")) {
				uint64_t total = 0;
				uint64_t current_index = 0;
				uint64_t current_amount = 0;
				plist_t current_list = NULL;
				instproxy_status_get_current_list(status, &total, &current_index, &current_amount, &current_list);
				if (current_list) {
					print_apps(current_list);
					plist_free(current_list);
				}
			} else {
				/* get progress if any */
				int percent = -1;
				instproxy_status_get_percent_complete(status, &percent);
                
                // FIX. Make sure areguments are not NULL
				if (last_status && status_name && (strcmp(last_status, status_name))) {
					printf("\r");
				}

				if (percent >= 0) {
					printf("%s: %s (%d%%)\n", command_name, status_name, percent);
				} else {
					printf("%s: %s\n", command_name, status_name);
				}
			}
		} else {
			/* report error to the user */
			if (error_description)
				fprintf(stderr, "ERROR: %s failed. Got error \"%s\" with code 0x%08"PRIx64": %s\n", command_name, error_name, error_code, error_description ? error_description: "N/A");
			else
				fprintf(stderr, "ERROR: %s failed. Got error \"%s\".\n", command_name, error_name);
			err_occurred = 1;
		}

		/* clean up */
		if (error_name)
			free(error_name);

		if (error_description)
			free(error_description);

		if (last_status) {
			free(last_status);
			last_status = NULL;
		}

		if (status_name) {
			last_status = strdup(status_name);
			free(status_name);
		}

		if (command_name) {
			free(command_name);
			command_name = NULL;
		}
	} else {
		fprintf(stderr, "ERROR: %s was called with invalid arguments!\n", __func__);
	}
}

**Comment by tvolkert** [comment]
I think @cbracken had a fix for this as well, but as far as I can tell, the libimobiledevice maintainers aren't responding to issues or pull requests...

**Comment by nikias** [comment]
Fixed with commit f07a9a0a4b844d12f70569ce8ca7c555ddb3cad8.
